zio/parquetio/iterator.go

```diff
@@ -0,0 +1,726 @@
+package parquetio
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+
+	"github.com/apache/thrift/lib/go/thrift"
+	"github.com/xitongsys/parquet-go/common"
+	"github.com/xitongsys/parquet-go/compress"
+	"github.com/xitongsys/parquet-go/layout"
+	"github.com/xitongsys/parquet-go/parquet"
+	"github.com/xitongsys/parquet-go/reader"
+	"github.com/xitongsys/parquet-go/source"
+)
+
+// Helper to avoid sprinkling fmt.Printf() calls around the code.
+// Enable debug statements by uncommenting the call to Printf()
+func debugf(msg string, args ...interface{}) {
+	//fmt.Printf(msg, args...)
+}
+
+// valueReader is the interface used for anything that can iterate over
+// a series of values of one of the parquet primitive types.  Note that
+// not all methods can be used on all implementations of this interface.
+type valueReader interface {
+	nextBoolean() bool
+	nextInt32() int32
+	nextInt64() int64
+	nextFloat() float64
+	nextDouble() float64
+	nextByteArray() []byte
+}
+
+// columnIterator reads and emits all the values from an individual
+// column inside a parquet file.  This implementation is not yet complete,
+// it only handles PLAIN and PLAIN_DICTIONARY encodings for a few
+// primitive data types.
+type columnIterator struct {
+	pr   *reader.ParquetReader
+	name string
+
+	maxRL int32
+	maxDL int32
+
+	rowGroup int
+
+	groupRead    int
+	groupTotal   int
+	thriftReader *thrift.TBufferedTransport
+	colMetadata  *parquet.ColumnMetaData
+
+	pageRead  int
+	pageTotal int
+
+	rlReader  *hybridReader
+	dlReader  *hybridReader
+	valReader valueReader
+
+	// only used for PageType_DICTIONARY_PAGE
+	int64Dict     []int64
+	floatDict     []float64
+	byteArrayDict [][]byte
+	// XXX need other types
+}
+
+func newColumnIterator(pr *reader.ParquetReader, name string, maxRL, maxDL int32) *columnIterator {
```

Comment:
In parquet-go do a `*source.ParquetFile.Open("")` to read each separate column. I think we might want to do this as well.

Reply:
That would be necessary if we had different goroutines reading different columns concurrently, but this code is all single-threaded.  I think an even better fix would be to pass in a go `io.ReaderAt` and then wrap that with something that we can give to `thrift.NewTBufferedTransport` so that we're not using a bunch of unneeded file descriptors...